Modules/_io/fileio.c

```diff
@@ -1141,16 +1148,16 @@ static PyObject *
 _io_FileIO_isatty_impl(fileio *self)
 /*[clinic end generated code: output=932c39924e9a8070 input=cd94ca1f5e95e843]*/
 {
-    long res;
-
     if (self->fd < 0)
         return err_closed();
-    Py_BEGIN_ALLOW_THREADS
-    _Py_BEGIN_SUPPRESS_IPH
-    res = isatty(self->fd);
-    _Py_END_SUPPRESS_IPH
-    Py_END_ALLOW_THREADS
-    return PyBool_FromLong(res);
+    if (self->isatty < 0) {
+        Py_BEGIN_ALLOW_THREADS
+        _Py_BEGIN_SUPPRESS_IPH
+        self->isatty = isatty(self->fd);
```

Comment:
There is an unsafe race condition here: other threads may be accessing `self->isatty` (or other fields) concurrent with this write. It's a problem even with the GIL enabled because of the `Py_BEGIN_ALLOW_THREADS`.